### PR TITLE
[Feature][NPU]: add runtime support for AutoRound quantized models

### PR DESCRIPTION
**Pull request by YChange01**
## Stacked PR Notice

> **This PR is dependent on #16364.**

Currently, the file changes include commits from PR #16364 (`feat(ascend NPU): add runtime support for GPTQ-quantized MoE models`) because that PR has not been merged yet.

**For Reviewers:**
1. Please **merge #16364 first**.
2. Once #16364 is merged, this PR will automatically update to show **only** the changes related to `auto_round.py`.
3. If you want to review now, please **ignore the GPTQ-related changes** and focus solely on:
   - File: `python/sglang/srt/layers/quantization/auto_round.py`

## PR Description

This PR adds runtime support for **AutoRound-quantized dense and MoE models** on the Ascend NPU backend.

### Key Changes
- Modified `python/sglang/srt/layers/quantization/auto_round.py` to enable Ascend compatibility.
- Aligned behavior with existing quantization logic to ensure safe handling of scales and offsets.

## Motivation

Currently, the Ascend runtime supports GPTQ-style quantization, but AutoRound
checkpoints could not be loaded or executed correctly.

Adding AutoRound support improves model compatibility and reduces manual
conversion overhead.

## Accuracy tests

Accuracy results and comparison logs are attached in the PR comments.

### CODE REVIEW

**Comment by cqchou** [comment]
## gsm8k

> python3 -m sglang.launch_server --model-path /root/.cache/Qwen3-30B-A3B --device npu --attention-backend ascend --host 0.0.0.0 --port 30000 --mem-fraction-static 0.8 --tp-size 2

```sh

python bench_sglang.py
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:30<00:00,  6.63it/s]
Accuracy: 0.940
Invalid: 0.000
Latency: 30.448 s
Output throughput: 747.104 token/s
```



> python3 -m sglang.launch_server --model-path /root/.cache/Qwen3-30B-A3B-GPTQ-Int4 --device npu --attention-backend ascend --host 0.0.0.0 --port 30000 --mem-fraction-static 0.8 --tp-size 2

```sh

python bench_sglang.py
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:36<00:00,  5.46it/s]
Accuracy: 0.910
Invalid: 0.000
Latency: 36.850 s
Output throughput: 685.396 token/s
```

File: python/sglang/srt/layers/quantization/auto_round.py [hunk 1]
```diff
@@ -301,6 +304,11 @@ def apply_awq_quant_layer(self, layer, prefix: str, backend: str = "auto"):
     def apply_gptq_quant_layer(self, layer, prefix: str, backend: str = "auto"):
         from sglang.srt.layers.linear import LinearBase
         from sglang.srt.layers.moe.fused_moe_triton import FusedMoE
+        from sglang.srt.layers.quantization.gptq import (
+            GPTQConfig,
+            GPTQLinearAscendMethod,
+            GPTQMoEAscendMethod,
+        )
         from sglang.srt.layers.quantization.marlin_utils import (
             check_marlin_supported,
```
Comment (ping1jing2) [inline python/sglang/srt/layers/quantization/auto_round.py:305]: actually, from my side, i don't think this is a good style, could you please move these import-statement to the top of this file.

File: python/sglang/srt/layers/quantization/gptq.py [hunk 2]
```diff
@@ -59,6 +62,10 @@
     )
 
 _is_cuda = is_cuda()
+_is_npu = is_npu()
+
+if _is_npu:
+    import torch_npu
```
Comment (ping1jing2) [inline python/sglang/srt/layers/quantization/gptq.py]: duplicate code, please remove it

File: python/sglang/srt/layers/quantization/gptq.py [hunk 3]
```diff
@@ -7,6 +7,9 @@
 
 import torch
 
+from sglang.srt.hardware_backend.npu.quantization.fused_moe_method_npu import (
```
Comment (ping1jing2) [inline python/sglang/srt/layers/quantization/gptq.py:10]: i think this file is mostly the same as PR #16364

File: python/sglang/srt/layers/linear.py [hunk 4]
```diff
@@ -60,6 +60,8 @@
     "GPTQMarlin24LinearMethod",
     "TPUInt8LinearMethod",
     "GPTQLinearMethod",
+    "GPTQLinearAscendMethod",
```
Comment (ping1jing2) [inline python/sglang/srt/layers/linear.py:67]: ditto

File: test/srt/ascend/test_ascend_autoround.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+from sglang.test.few_shot_gsm8k import run_eval as run_eval_few_shot_gsm8k
+from sglang.test.test_utils import (
+    DEFAULT_TIMEOUT_FOR_SERVER_LAUNCH,
+    DEFAULT_URL_FOR_TEST,
+    CustomTestCase,
+    popen_launch_server,
+)
+
+TEST_MODEL_MATRIX = {
+    "Qwen/Qwen3-30B-A3B-Instruct-2507-int4-AutoRound": {
```
Comment (ping1jing2) [inline test/srt/ascend/test_ascend_autoround.py]: same problems as PR #16364

File: test/srt/ascend/test_ascend_autoround.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            "--attention-backend",
+            "ascend",
+            "--quantization",
+            "auto-round",
+        ]
+
+    def test_a_gsm8k(self):
+        for model in self.models:
+            with self.subTest(model=model):
+                print(f"##=== Testing accuracy: {model} ===##")
```
Comment (ping1jing2) [inline test/srt/ascend/test_ascend_autoround.py]: please use `logger` instead of `print`

File: test/srt/ascend/test_ascend_gptq_moe.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
+import unittest
+from types import SimpleNamespace
+from urllib.parse import urlparse
+
+from sglang.srt.utils import kill_process_tree
```
Comment (ping1jing2) [inline test/srt/ascend/test_ascend_gptq_moe.py:6]: ditto

File: python/sglang/srt/layers/quantization/auto_round.py (hunk 1 above)
Comment (YChange01) [reply python/sglang/srt/layers/quantization/auto_round.py:305]: This coding style is consistent with the rest of this file, where local imports are used extensively. I suggest we keep this PR focused on the current fix and refactor the import style for the whole file in a separate PR.

File: python/sglang/srt/layers/linear.py (hunk 4 above)
Comment (YChange01) [reply python/sglang/srt/layers/linear.py:67]: fixed

File: test/srt/ascend/test_ascend_autoround.py (hunk 6 above)
Comment (YChange01) [reply test/srt/ascend/test_ascend_autoround.py]: fixed

File: test/srt/ascend/test_ascend_gptq_moe.py (hunk 7 above)
Comment (YChange01) [reply test/srt/ascend/test_ascend_gptq_moe.py:6]: fixed

File: python/sglang/srt/models/qwen3.py [hunk 8]
```diff
@@ -161,12 +161,12 @@ def forward_prepare_npu(self, positions, hidden_states, forward_batch):
...
             self.rotary_emb.position_sin,
             self.rotary_emb.position_cos,
+            q_weight=self.q_norm.weight,
+            k_weight=self.k_norm.weight,
             self.q_size,
             self.kv_size,
             self.head_dim,
             eps=self.q_norm.variance_epsilon,
-            q_weight=self.q_norm.weight,
-            k_weight=self.k_norm.weight,
```
Comment (TamirBaydasov) [inline python/sglang/srt/models/qwen3.py]: https://github.com/sgl-project/sglang/pull/16364#discussion_r2821706764

File: python/sglang/srt/models/qwen3_moe.py [hunk 9]
```diff
@@ -530,12 +530,12 @@ def forward_prepare_npu(
...
             self.rotary_emb.position_sin,
             self.rotary_emb.position_cos,
+            self.q_norm.weight,
+            self.k_norm.weight,
             self.q_size,
             self.kv_size,
             self.head_dim,
             eps=self.q_norm.variance_epsilon,
-            q_weight=self.q_norm.weight,
-            k_weight=self.k_norm.weight,
```
Comment (TamirBaydasov) [inline python/sglang/srt/models/qwen3_moe.py:538]: https://github.com/sgl-project/sglang/pull/16364#discussion_r2821706764

File: python/sglang/srt/models/qwen3.py (hunk 8 above)
Comment (YChange01) [reply python/sglang/srt/models/qwen3.py]: After updating sgl-kernel-npu to the latest version, the modifications in python/sglang/srt/models/qwen3_moe.py are no longer required. I have reverted this file to align with the main branch and updated the PR accordingly.

File: python/sglang/srt/models/qwen3_moe.py (hunk 9 above)
Comment (YChange01) [reply python/sglang/srt/models/qwen3_moe.py:538]: ditto
